### PR TITLE
Fix: Mongoose types incorrect for when includeResultMetadata: true is set

### PR DESCRIPTION
**Pull request by prathamVaidya**
**Summary**

Fixed Issue #13987 
Mongoose Version: 8.0

When `includeResultMetadata` is set to `true`. The document parameter in post middleware can also be `ModifyResult<>`.

Functions that support `includeResultMetadata` option are : `findOneAndUpdate`, `findByIdAndUpdate`, `findOneAndReplace`, `findByIdAndDelete`, and `findOneAndDelete`.

Changes Done:

- [x] Introduced a separate document type for the post middleware, specifically for these functions, ensuring they also return `ModifyResult<>`.
- [x] Some of these functions were not returning `ModifyResult<>` when `includeResultMetadata` was set to True. I've addressed this by adding appropriate return types.
- [x] In cases where `includeResultMetadata` is false and no matching document is found in any of these functions, the document parameter in the post middleware is `NULL`. So I have included `NULL` as a possible doc type in the middleware parameter.

**Examples**

```ts
import mongoose, { ModifyResult, Query } from "mongoose";

const UserSchema = new mongoose.Schema({
  id: String,
  username: String,
});

UserSchema.post<
  Query<
    mongoose.HydratedDocument<mongoose.InferSchemaType<typeof UserSchema>>,
    mongoose.InferSchemaType<typeof UserSchema>
  >
>("findOneAndDelete", function (doc) {
// doc can be NULL | QueryResultType<T> | ModifyResult<QueryResultType<T>>
  if (doc && "value" in doc) {
    console.debug("DELETE VALUE = ", doc.value);
  }
});

const UserModel = mongoose.model("User", UserSchema);

async function main() {
  await mongoose.connect(`mongodb://localhost:27017/mongoose-test`);

  await mongoose.connection.dropDatabase();

  const doc = new UserModel({ username: "user1" });

  await doc.save();

  const upsertResult = await UserModel.findOneAndDelete(
    {username: "user1"},
    {
      includeResultMetadata: true,
    }
  ).exec();

  if (upsertResult) {
    console.log("Result=", upsertResult.value); // it has type set to ModifyResult now
  } else {
    console.log("NULL", upsertResult);
  }
}

main();

```

I have fixed the issue for v8.0. If the solution is acceptable then I will also do it for v7.4 that was originally specified in the Issue. @hasezoey

### CODE REVIEW

File: types/models.d.ts [hunk 1]
```diff
@@ -565,6 +569,10 @@ declare module 'mongoose' {
       TRawDocType,
       'findOneAndDelete'
     >;
+    findByIdAndRemove<ResultDoc = THydratedDocumentType>(
```
Comment (vkarpov15) [inline types/models.d.ts:572]: `findByIdAndRemove()` is no longer supported in Mongoose 8, so please remove this

File: types/middlewares.d.ts [hunk 2]
```diff
@@ -5,6 +5,7 @@ declare module 'mongoose' {
   type MongooseDistinctDocumentMiddleware = 'save' | 'init' | 'validate';
   type MongooseDocumentMiddleware = MongooseDistinctDocumentMiddleware | MongooseQueryAndDocumentMiddleware;
 
+  type MongooseRawResultQueryMiddleware = 'findOneAndUpdate' | 'findOneAndReplace' | 'findOneAndDelete' | 'findByIdAndUpdate' | 'findOneAndDelete';
```
Comment (vkarpov15) [inline types/middlewares.d.ts:8]: `findByIdAndUpdate` doesn't have distinct middleware, that function fires `findOneAndUpdate` middleware

Comment (vkarpov15) [reply types/middlewares.d.ts:8]: Also it looks like you have `findOneAndDelete` twice here

File: types/models.d.ts [hunk 3]
```diff
@@ -640,11 +647,6 @@ declare module 'mongoose' {
       replacement: TRawDocType | AnyObject,
       options: QueryOptions<TRawDocType> & { includeResultMetadata: true }
     ): QueryWithHelpers<ModifyResult<ResultDoc>, ResultDoc, TQueryHelpers, TRawDocType, 'findOneAndReplace'>;
-    findOneAndReplace<ResultDoc = THydratedDocumentType>(
```
Comment (vkarpov15) [inline types/models.d.ts:643]: Why did you remove `findOneAndReplace` here?

**Review by vkarpov15 — CHANGES_REQUESTED** [review]
Thanks for your PR. Please fix lint, add a test case, and address the comments in this review :+1:

File: types/models.d.ts (hunk 3 above)
Comment (prathamVaidya) [reply types/models.d.ts:643]: There was a duplicate of findOneAndReplace with option ` { includeResultMetadata: true }` .
It was from old commit where someone replaced `rawResult` option to  `includeResultMetadata` which created duplicates.

I found 2 occurrences of this code. So removed one. Also removed duplicate for `findByIdAndUpdate`. 

```
findOneAndReplace<ResultDoc = THydratedDocumentType>(
      filter: FilterQuery<TRawDocType>,
      replacement: TRawDocType | AnyObject,
      options: QueryOptions<TRawDocType> & { includeResultMetadata: true }
```

File: types/middlewares.d.ts (hunk 2 above)
Comment (prathamVaidya) [reply types/middlewares.d.ts:8]: Thanks @vkarpov15,  I'll update.

File: types/models.d.ts (hunk 1 above)
Comment (prathamVaidya) [reply types/models.d.ts:572]: Ok, should I delete all the `findByIdAndRemove` type definitions?

Comment (prathamVaidya) [reply types/models.d.ts]: I have delete all type definations for `findByIdAndRemove`

File: types/middlewares.d.ts [hunk 4]
```diff
@@ -5,6 +5,7 @@ declare module 'mongoose' {
   type MongooseDistinctDocumentMiddleware = 'save' | 'init' | 'validate';
   type MongooseDocumentMiddleware = MongooseDistinctDocumentMiddleware | MongooseQueryAndDocumentMiddleware;
 
+  type MongooseRawResultQueryMiddleware = 'findOneAndUpdate' | 'findOneAndReplace' | 'findOneAndDelete';
   type MongooseDistinctQueryMiddleware = 'count' | 'estimatedDocumentCount' | 'countDocuments' | 'deleteMany' | 'distinct' | 'find' | 'findOne' | 'findOneAndDelete' | 'findOneAndReplace' | 'findOneAndUpdate' | 'replaceOne' | 'updateMany';
```
Comment (hasezoey) [inline types/middlewares.d.ts:9]: shouldnt `MongooseDistinctQueryMiddleware` inherit `MongooseRawResultQueryMiddleware` instead of defining it duplicately?

File: types/models.d.ts [hunk 5]
```diff
@@ -587,11 +575,6 @@ declare module 'mongoose' {
       update: UpdateQuery<TRawDocType>,
       options: QueryOptions<TRawDocType> & { includeResultMetadata: true }
     ): QueryWithHelpers<ModifyResult<ResultDoc>, ResultDoc, TQueryHelpers, TRawDocType, 'findOneAndUpdate'>;
-    findByIdAndUpdate<ResultDoc = THydratedDocumentType>(
-      id: mongodb.ObjectId | any,
-      update: UpdateQuery<TRawDocType>,
-      options: QueryOptions<TRawDocType> & { includeResultMetadata: true }
-    ): QueryWithHelpers<ModifyResult<ResultDoc>, ResultDoc, TQueryHelpers, TRawDocType, 'findOneAndUpdate'>;
```
Comment (hasezoey) [inline types/models.d.ts:594]: was this also removed because of a duplicate?

**Review by hasezoey — APPROVED** [review]
Looks good to me, though some minor questions / suggestions.

also does someone know why the tests are not running for the latest commits?

File: types/middlewares.d.ts (hunk 4 above)
Comment (prathamVaidya) [reply types/middlewares.d.ts:9]: While `MongooseDistinctQueryMiddleware` is a superset of `MongooseRawResultQueryMiddleware`. I don't think its a good idea to inherit it because in case where we create a new type for middlewares that support some other option than that type can also contain middlewares from `MongooseRawResultQueryMiddleware` and vice versa. According to me, It's better if we maintain a separate type for this. Let me know your final decision and I'll update accordingly.

File: types/models.d.ts (hunk 5 above)
Comment (prathamVaidya) [reply types/models.d.ts:594]: Yes, this was also duplicate. I removed duplicates for these two: `findByIdAndUpdate` and `findOneAndReplace`

File: types/middlewares.d.ts (hunk 4 above)
Comment (hasezoey) [reply types/middlewares.d.ts:9]: i will be fine with it either way, @vkarpov15 what do you think?

Comment (prathamVaidya) [reply types/middlewares.d.ts:9]: @hasezoey, once all the changes are completed and it's ready to be approved, there are conflicts of the branch merging process due to the latest pushes. I just wanted to know, do I have to resolve these conflicts, or will the reviewer handle it?

Comment (vkarpov15) [reply types/middlewares.d.ts:9]: I think avoiding inheriting is a reasonable idea.

**Comment by vkarpov15** [comment]
Can you please fix merge conflicts?

**Comment by prathamVaidya** [comment]
Fixed Merge Conflicts ✅

**Review by vkarpov15 — APPROVED** [review]
LGTM, thanks :+1:
